### PR TITLE
Docs: add isolated systems guide

### PR DESCRIPTION
**Pull request by treysp**
Also adds the mkdocs-glightbox extension so images expand/zoom when they are clicked

### CODE REVIEW

File: docs/guides/isolated_systems.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
+# Isolated systems guide
+
+SQLMesh assumes that production and non-production data systems can access one another, but that is not always the case.
```
Comment (tobymao) [inline docs/guides/isolated_systems.md:3]: what about something like sqlmesh is optimized for use cases where developers have access to production data

File: docs/guides/isolated_systems.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+Some architectures isolate systems from one another, preventing production data from being accessed in non-production systems. This isolation is usually motivated by information security concerns, as non-production systems may be accessible to more users and/or have looser security restrictions.
+
+This guide explains how to use SQLMesh with isolated systems. It also clarifies how isolating systems affects SQLMesh's behavior and why the systems still enjoy the benefits of many SQLMesh features.
+
+## Terminology
+
+Isolated systems are sometimes referred to as "isolated environments," but that term is ambiguous.
+
+In this context, the things being isolated are servers and databases, and the isolation is enforced by limiting network communications between the systems. We will refer to these as "isolated systems" - the "production system" and "non-production system."
```
Comment (tobymao) [inline docs/guides/isolated_systems.md:13]: this reads like a technical paper :). can we make it a bit more relatable like two separate data warehouses that can't talk to each other?

File: docs/guides/isolated_systems.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+When we refer to "environments," we are always talking about [SQLMesh environments](../concepts/environments.md) - the isolated namespaces created and managed by SQLMesh.
+
+## Configuring SQLMesh
+
+### Separate state data
+
+SQLMesh maintains a record of every model version, to which it compares updated models. The records are called "state" data, as in "the state of the model at that point in time."
+
+State data can be stored alongside model data in the primary data warehouse or in a [separate database](./connections.md#state-connection). (We recommend using a separate transactional database for projects running on cloud SQL engines.)
```
Comment (tobymao) [inline docs/guides/isolated_systems.md:23]: did you define model data?

File: docs/guides/isolated_systems.md [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+This section outlines a generic workflow for SQLMesh projects with one system, which serves as a baseline to which we will compare the workflow for isolated systems.
+
+We assume that a version of the SQLMesh project is currently running in production and serves as the starting point for code modifications.
+
+The simplest version of the workflow is:
+
+1. Developers modify project code
+
+    a. Each developer makes their own copy of the production SQLMesh project files
```
Comment (tobymao) [inline docs/guides/isolated_systems.md:99]: maybe just say use git clone? or use git terms?

File: docs/guides/isolated_systems.md [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+We now describe the SQLMesh project workflow with isolated systems.
+
+The primary difference from the single-system workflows described above is that some steps now occur in the nonproduction system and others in the production system.
+
+In some isolated system implementations, the data in the non-production system is different from that in the production system. For example, columns containing sensitive information may be masked or altered in the non-production system's data. These potential differences increase the likelihood of unintentional errors, so we describe (and recommend) a workflow with staging environments.
+
+In the steps below, we assume that the non-production system is the default gateway. The isolated system workflow with staging environments is:
+
+1. Developers modify project code in the non-production environment
```
Comment (z3z1ma) [inline docs/guides/isolated_systems.md:163]: Wouldnt it be simpler to just call this cloning the git repository. And if not, wouldn't it be simpler to just say that the developer workflow is identical to the prior section barring the fact you should specify the non-prod gateway in the CLI if its not the default. This makes it seem like a very different workflow and the differing verbiage is confusing.

File: docs/guides/isolated_systems.md [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+    b. Developers update code in their copy of the project files
+
+    c. Developers create SQLMesh environments and validate their code updates
+
+2. Update production project code
+
+    a. A set of code updates is selected for deployment to production
+
+    b. The production SQLMesh project files are modified with the updates
```
Comment (z3z1ma) [inline docs/guides/isolated_systems.md:139]: Werent they already modified in step 1b

File: docs/guides/isolated_systems.md [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+3. Deploy to production system `staging` environment
+
+    a. Execute the `sqlmesh --gateway production plan staging` command to apply the updates to the production system `staging` environment. Note that your production system's gateway may have a different name and that you may choose any environment name other than `prod`.
+
+    b. Developers validate the set of code updates in the production system `staging` environment
+
+4. Deploy to production
+
+    a. Execute the `sqlmesh  --gateway production plan` command to apply the updates to the production system's `prod` environment
```
Comment (z3z1ma) [inline docs/guides/isolated_systems.md:189]: The classifications of breaking/non-breaking are lost across gateways. This must be called out. Reconciliation requires auto classification which is subpar, but apparently the most common. Or use of the github CICD bot to perform interactive classification in the PR.

Comment (treysp) [reply docs/guides/isolated_systems.md:167]: Added info on this on line 171 below - is that sufficient?

File: docs/guides/isolated_systems.md (hunk 4 above)
Comment (treysp) [reply docs/guides/isolated_systems.md]: Updated

File: docs/guides/isolated_systems.md (hunk 5 above)
Comment (treysp) [reply docs/guides/isolated_systems.md]: Agreed - do the updates make the similarity more apparent?

File: docs/guides/isolated_systems.md (hunk 7 above)
Comment (z3z1ma) [reply docs/guides/isolated_systems.md:167]: Yeah!

File: docs/concepts/macros/macro_variables.md [hunk 8]
```diff
@@ -93,10 +95,20 @@ All predefined macro variables:
...
 
-Other macro variables:
+### Runtime variables
+
+SQLMesh provides two other predefined variables used to modify model behavior based on information available at runtime.
 
 * @runtime_stage - A string value that denotes the current stage of the SQLMesh runtime. It can take one of the following values:
     * 'loading' - The project is currently being loaded into SQLMesh's runtime context.
     * 'creating' - The model tables are being created.
     * 'evaluating' - The models' logic is being evaluated.
```
Comment (georgesittas) [inline docs/concepts/macros/macro_variables.md:105]: Forgot to update this in https://github.com/TobikoData/sqlmesh/commit/adcd016784e3c89d13d0cd8ec15135195106977d

```suggestion
    * 'evaluating' - The models' logic is being evaluated.
    * 'testing' - The models' logic is being evaluated in the context of a unit test.
```

File: docs/guides/isolated_systems.md [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+
+In the steps below, we assume that the non-production system is the default gateway. The isolated system workflow with staging environments begins like the [staging workflow above](#staging-environment), but with additional staging environment creation and validation in the production system after validation in the default non-production system:
+
+1. [Same as [above](#staging-environment)]
+
+2. [Same as [above](#staging-environment)]
+
+3. Deploy to production system `staging` environment
+
+    a. Execute the `sqlmesh --gateway production plan staging` command to apply the updates to the production system `staging` environment. Note that your production system's gateway may have a different name and that you may choose any environment name other than `prod`.
```
Comment (georgesittas) [inline docs/guides/isolated_systems.md:161]: Is this a typo?
```suggestion
    a. Execute the `sqlmesh --gateway production plan staging` command to apply the updates to the production system `staging` environment. Note that your production system's gateway may have a different name and that you may choose any environment name other than `staging`.
```

File: docs/guides/isolated_systems.md [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+
+In isolated systems, SQLMesh's virtual data environments will operate normally *within* each system, but not across systems. For example, when the non-production system's staging environment is created (step 2c), any computations previously run in a developer's personal environment will not run again.
+
+In step (3) above, the creation of the `staging` environment in the production system will require running all the computations required by code changes. Deployment to the production system's `prod` environment (step 4) will not require additional computation due to SQLMesh's virtual data environments. The deployment only requires updates views so will be very rapid.
+
+## SQLMesh features
+
+This section describes how two key SQLMesh features perform when a project runs on isolated systems: virtual data environments and computing only what's necessary.
+
+### Virtual data environments
```
Comment (georgesittas) [inline docs/guides/isolated_systems.md:181]: Do we already have a doc page for VDEs that we could reuse? This seems something which may be better off documented in its own page and then referenced here, as it's a core SQLMesh concept.

File: docs/guides/isolated_systems.md [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+
+When an existing model is modified, a new physical table is created to reflect the updated model. The virtual layer of any environment containing that version of the model can point to the same new physical table. This approach allows a model version's computations to run once and be reused by any environment (including production).
+
+#### Table reuse in isolated systems
+
+Isolated production systems cannot reuse the physical tables created in non-production systems. Therefore, computations performed in non-production systems must be run again when an updated model version is pushed to the production system.
+
+However, computations performed in different SQLMesh environments inside a system will be reused by other SQLMesh environments in that system. In organizations with many collaborators, most computations occur during development, resulting in time and cost savings due to table reuse within the non-production system.
+
+### Compute only what's necessary
```
Comment (georgesittas) [inline docs/guides/isolated_systems.md:197]: These last two sections felt a little redundant, but I may be biased. After reading the whole thing, it seemed to me that it'd already been explained how SQLMesh only computes what's necessary, eg. L117, L143, L173, L195.

File: docs/guides/isolated_systems.md [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+When we refer to "environments," we are always talking about [SQLMesh environments](../concepts/environments.md) - the isolated namespaces created and managed by SQLMesh.
+
+## Configuring SQLMesh
+
+### Separate state data
+
+SQLMesh maintains a record of every model version so it can identify changes when models are updated. Those records are called "state" data, as in "the state of the model at that point in time."
+
+State data can be stored alongside other data in the primary data warehouse or in a [separate database](./connections.md#state-connection). (We recommend using a separate transactional database for projects running on cloud SQL engines.)
```
Comment (eakmanrq) [inline docs/guides/isolated_systems.md:25]: Would this be better to link to? https://sqlmesh.readthedocs.io/en/stable/guides/configuration/#state-connection

Or maybe some of that information should be copied over?

File: docs/guides/isolated_systems/isolated-systems_linkage.png
Comment (eakmanrq) [inline docs/guides/isolated_systems/isolated-systems_linkage.png:1]: Should the digram have project files that then point to the two systems? Right now to me it reads like "Non-production system loads project files which then is used in the production system".

**Review by eakmanrq — APPROVED** [review]
Great to have a guide for this now! 👏

File: docs/guides/isolated_systems/isolated-systems_linkage.png
Comment (treysp) [reply docs/guides/isolated_systems/isolated-systems_linkage.png:1]: good call

**Pull request merged**